src/backend/distributed/utils/causal_clock.c

```diff
@@ -0,0 +1,669 @@
+
+	LWLockRelease(&LogicalClockShmem->clockLock);
+
+	ereport(DEBUG1, (errmsg("adjusted to remote or persisted clock: "
+							"<logical(%lu) counter(%u)>",
+							remoteClock->logical,
+							remoteClock->counter)));
+}
+
+
+/*
+ * GetHighestClockInTransaction takes the connection list of participating nodes in the
+ * current transaction and polls the logical clock value of all the nodes. Returns the
+ * highest logical clock value of all the nodes in the current distributed transaction,
+ * which may be used as commit order for individual objects in the transaction.
+ */
+static ClusterClock *
+GetHighestClockInTransaction(List *nodeConnectionList)
+{
+	MultiConnection *connection = NULL;
+
+	foreach_ptr(connection, nodeConnectionList)
+	{
+		int querySent =
+			SendRemoteCommand(connection, "SELECT citus_get_cluster_clock();");
+
+		if (querySent == 0)
+		{
+			ReportConnectionError(connection, ERROR);
+		}
+	}
+
+	ClusterClock *globalClockValue = NULL;
+
+	/* fetch the results and pick the highest clock value of all the nodes */
+	foreach_ptr(connection, nodeConnectionList)
+	{
+		bool raiseInterrupts = true;
+
+		if (PQstatus(connection->pgConn) != CONNECTION_OK)
+		{
+			ereport(ERROR, (errmsg("connection to %s:%d failed when "
+								   "fetching logical clock value",
+								   connection->hostname, connection->port)));
+		}
+
+		PGresult *result = GetRemoteCommandResult(connection, raiseInterrupts);
+		if (!IsResponseOK(result))
+		{
+			ReportResultError(connection, result, ERROR);
+		}
+
+		ClusterClock *nodeClockValue = ParseClusterClockPGresult(result, 0, 0);
+
+		ereport(DEBUG1, (errmsg("node(%lu:%u) transaction clock %lu:%u",
+								connection->connectionId, connection->port,
+								nodeClockValue->logical, nodeClockValue->counter)));
+
+		if (globalClockValue)
+		{
+			globalClockValue = LargerClock(globalClockValue, nodeClockValue);
+		}
+		else
+		{
+			globalClockValue = nodeClockValue;
+		}
+
+		PQclear(result);
+		ForgetResults(connection);
+	}
+
+	ereport(DEBUG1, (errmsg("final global transaction clock %lu:%u",
+							globalClockValue->logical,
+							globalClockValue->counter)));
+	return globalClockValue;
+}
+
+
+/*
+ * AdjustClocksToTransactionHighest Sets the clock value of all the nodes, participated
+ * in the PREPARE of the transaction, to the highest clock value of all the nodes.
+ */
+static void
+AdjustClocksToTransactionHighest(List *nodeConnectionList, ClusterClock
+								 *transactionClockValue)
+{
+	StringInfo queryToSend = makeStringInfo();
+
+	/* Set the adjusted value locally */
+	AdjustLocalClock(transactionClockValue);
+
+	/* Set the clock value on participating worker nodes */
+	appendStringInfo(queryToSend,
+					 "SELECT pg_catalog.citus_internal_adjust_local_clock_to_remote"
+					 "('(%lu, %u)'::pg_catalog.cluster_clock);",
+					 transactionClockValue->logical, transactionClockValue->counter);
+
+	ExecuteRemoteCommandInConnectionList(nodeConnectionList, queryToSend->data);
```

Comment:
this is going to happen outside of the current coordinated transaction. Given that we use sequences, that should be fine. Is that right?

Reply:
You are right, sequences are non-transactional, but aren't these connections(InProgressTransactions) are part of the current transaction?